### PR TITLE
Provided parsing

### PR DESCRIPTION
**Pull request by BrianEdwards85**
Testing will be added, wanted to provided the working parser for review while I write tests

### CODE REVIEW

**Comment by BrianEdwards85** [comment]
not ready to be merged, it functions as expected, but no tests yet.  Going to be writing tests around the new code this afternoon, but wanted to get some eyes on the  new code

File: src/smidje/parser.clj [hunk 1]
```diff
@@ -15,11 +20,77 @@
   (and (> (count input) 3)
        (provided-form? (nth input 3))))
 
+(defn- gen-provided-sym
+  [fn1 fn2]
+  (symbol (str ".." (gensym (str fn1 "->" fn2)) ".." ))) 
+
+(defn- flatten-provided
```
Comment (munk) [inline src/smidje/parser.clj]: could the flatten be omitted here and use the flatten function?

File: src/smidje/parser.clj [hunk 2]
```diff
@@ -15,11 +20,74 @@
...
+  [provided]
+  (let [match (first provided)]
+    (if (some list? match)
```
Comment (munk) [inline src/smidje/parser.clj]: could this be `when`?

File: src/smidje/parser.clj (hunk 1 above)
Comment (BrianEdwards85) [inline src/smidje/parser.clj]: Would it be clearer to rename this unnest-provided ?

File: src/smidje/parser.clj [hunk 3]
```diff
@@ -15,11 +20,74 @@
...
+        (loop [provided-flattened [provided-fn] flattened-paramaters [] r (rest match)]
+          (cond
+            (empty? r) (conj flattened-paramaters (into []  (conj (rest provided ) (flatten  provided-flattened))))
+
+            (and (list? (first r))
+                 (not-empty (first r)))
+               (let [sub-provide (first r)
+                     sub-fn (first sub-provide)
+                     metaconst (gen-provided-sym provided-fn sub-fn)]
+                 (recur (conj provided-flattened metaconst) (into flattened-paramaters (flatten-provided [sub-provide '=> metaconst])) (rest r)))
```
Comment (munk) [inline src/smidje/parser.clj]: the loop and recursion is pretty hairy

File: src/smidje/parser.clj (hunk 1 above)
Comment (munk) [inline src/smidje/parser.clj]: Yeah, I like that

File: src/smidje/parser.clj (hunk 2 above)
Comment (munk) [inline src/smidje/parser.clj]: Oh, I misread this. nvm

File: src/smidje/parser.clj [hunk 4]
```diff
@@ -15,11 +20,74 @@
...
+  (symbol (str ".." (gensym (str fn1 "->" fn2)) ".." ))) 
+
+(defn- unnest-provided
+  [provided]
+  (let [match (first provided)]
+    (if (some list? match)
+      (let [provided-fn (first match)]
+        (loop [provided-flattened [provided-fn] flattened-paramaters [] r (rest match)]
+          (cond
+            (empty? r) (conj flattened-paramaters (into []  (conj (rest provided ) (flatten  provided-flattened))))
```
Comment (dustinvogt) [inline src/smidje/parser.clj]: Nit: extra space in (rest provided )`

**Pull request merged**